src/gt4py/next/iterator/ir_utils/common_pattern_matcher.py

```diff
@@ -7,13 +7,52 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 from collections.abc import Iterable
-from typing import Any, TypeGuard
+from typing import Any, Generic, List, TypeAlias, TypeGuard, TypeVar
 
 from gt4py.next.iterator import ir as itir
 from gt4py.next.iterator.ir_utils import ir_makers as im
 
 
-def is_applied_lift(arg: itir.Node) -> TypeGuard[itir.FunCall]:
+_Fun = TypeVar("_Fun", bound=itir.Expr)
+
+
+class _IsCall(itir.FunCall, Generic[_Fun]):
+    fun: _Fun
+    args: List[itir.Expr]
+
+
+_IsCallToPattern: TypeAlias = _IsCall[itir.SymRef]
```

Comment:
It's not a blocker but I think these names can be improved. What about `_IsCall` -> `FunCallTo` , `_IsCallToPattern` -> `FunCallToSymRef`, and `_IsFunCallToFunCallToRef` -> `FunCallToFunCallToSymRef` ?